### PR TITLE
Server WMS getPrint: fix HIGHLIGHT with follow map themes

### LINKED ISSUE #34178: QGIS Server GetPrint: HIGHLIGHT_GEOM is not printed if map layers are configured to follow a map theme
**Describe the bug**

The `HIGHLIGHT_GEOM` is handy to print some additional geometry on top of an existing layout.

If the user chooses to "follow a map theme", in the map properties, the additional geometry is not printed. The layers printed are the ones on the theme, as expected, but not the `HIGHLIGHT_GEOM`.

follow map theme

In my use case, I want to provide always the same layers to be printed (a set of official layers) plus the geometry (the parcel) edited by the user on the web.

This can be considered a feature request.

**How to Reproduce**

Using two almost identical layouts (one the "follow a map theme" set and the other not) and issue a GetPrint request with an additional geometry.

**QGIS and OS versions**

QGIS Master

### PR DESCRIPTION
**Pull request by elpaso**
Fixes #34178

### CODE REVIEW

**Comment by elpaso** [comment]
@nyalldawson I thought that `* Instead, use QgsLayoutItemMap::setLayers() to control which layers are rendered.` meant that by calling that method I could override the `mapThemeCollection()->mapThemeVisibleLayers` logic but there is something else going on here.

Because there is an `else if` the two are of course mutually exclusive, I think that the failing test in `TestQgsLayoutMap` is a false positive, the client code should not call `setLayers` on the map layout item if the map theme export is in control.

The main change is https://github.com/qgis/QGIS/pull/43391/files#diff-fad5b53afb20f316460160db51ff9c26f65d1108d00843ef18707c03c22d0b43R2134 where I inverted the priority of layers and themes effectively allowing client code to set layer and override the themes when needed.

**Comment by nyalldawson** [comment]
> I thought that * Instead, use QgsLayoutItemMap::setLayers() to control which layers are rendered. meant that by calling that method I could override the mapThemeCollection()->mapThemeVisibleLayers logic but there is something else going on here.

QgsLayoutItemMap::keepLayerSet / setKeepLayerSet is basically nonsense API -- it's used to turn on or off a checkbox in the layout map widget only. That's it :)

If a preset name is set then that overrules everything (as you've found)

Could the changes you've made here be avoided by turning off the map theme in the server code and then setting the layers to match the layers from the desired theme?

**Comment by elpaso** [comment]
> > I thought that * Instead, use QgsLayoutItemMap::setLayers() to control which layers are rendered. meant that by calling that method I could override the mapThemeCollection()->mapThemeVisibleLayers logic but there is something else going on here.
> 
> QgsLayoutItemMap::keepLayerSet / setKeepLayerSet is basically nonsense API -- it's used to turn on or off a checkbox in the layout map widget only. That's it :)
> 
> If a preset name is set then that overrules everything (as you've found)
> 
> Could the changes you've made here be avoided by turning off the map theme in the server code and then setting the layers to match the layers from the desired theme?

I've considered that approach first, but what do you mean with "turning off"? Because the only thing that came to my mind was the approach taken by @jgrocha in his PR (with a restorer) but I was looking for a simpler approach, the fact is that I cannot change the map layout object on each request because of the stateless nature of the server and because the project instance is cached and reused on future request.

I believe that we should implement the following mutually exclusive logic with the priority indicated in `QgsLayoutItemMap::layersToRender()`:
1. client code provide a list of layers to be rendered with `setLayers()`
2. if follow map theme is set the list of layers  is taken from the map theme ThemeCollection()
3. visible layers and order are taken from the project's layer tree

This would allow client code to pre-configure visible layersby calling `setLayers()`.

I also considered adding a flag to the map object to ignore the theme and leave the original logic in the method untouched but I'm convinced that the a.m. logic is what we really need without requiring API changed.

**Comment by elpaso** [comment]
> > I thought that * Instead, use QgsLayoutItemMap::setLayers() to control which layers are rendered. meant that by calling that method I could override the mapThemeCollection()->mapThemeVisibleLayers logic but there is something else going on here.
> 
> QgsLayoutItemMap::keepLayerSet / setKeepLayerSet is basically nonsense API -- it's used to turn on or off a checkbox in the layout map widget only. That's it :)
> 
> If a preset name is set then that overrules everything (as you've found)
> 
> Could the changes you've made here be avoided by turning off the map theme in the server code and then setting the layers to match the layers from the desired theme?

Ok, I didn't realize that we are working on a cloned template so this approach might work, testing it now.

**Comment by elpaso** [comment]
@jgrocha this should fix your issue, I wrote a whole new server test for this case but I'd like you to test it.

File: python/core/auto_generated/layout/qgslayoutitemmap.sip.in [hunk 1]
```diff
@@ -508,12 +508,15 @@ be used instead.
 
     void setLayers( const QList<QgsMapLayer *> &layers );
 %Docstring
-Sets the stored ``layers`` set. If empty, the current project layers will
-be used instead.
+Sets the stored ``layers`` set. If empty, the current project layers or the
+layer specified in the map theme (if the follow visibility preset flag is set)
```
Comment (nyalldawson) [inline python/core/auto_generated/layout/qgslayoutitemmap.sip.in:512]: This change isn't correct, is it?

Comment (elpaso) [reply python/core/auto_generated/layout/qgslayoutitemmap.sip.in:512]: Here's the implementation: https://github.com/qgis/QGIS/blob/master/src/core/layout/qgslayoutitemmap.cpp#L2130

I just wanted to point it out what are the options here. How would you suggest to improve this text?

**Comment by jgrocha** [comment]
> @jgrocha this should fix your issue, I wrote a whole new server test for this case but I'd like you to test it.

Yabadabadoo! Using the same project used for the bug report, the issue can not be reproduced anymore. Great work.

You have found a nice approach with a minimal change to the core QGIS. Doing it strictly on the server side was much more painful. 

The test added is very good. Probably writing the test took longer than the fix.

Thanks @elpaso

**Comment by rldhont** [comment]
Is this PR needs to be backported ?

**Comment by elpaso** [comment]
> Is this PR needs to be backported ?

Maybe :)  Do you feel it's safe?

**Comment by elpaso** [comment]
> > @jgrocha this should fix your issue, I wrote a whole new server test for this case but I'd like you to test it.
> 
> Yabadabadoo! Using the same project used for the bug report, the issue can not be reproduced anymore. Great work.
> 
> You have found a nice approach with a minimal change to the core QGIS. Doing it strictly on the server side was much more painful.

Actually, this is a  server-only change (except for the small method documentation change), what do you mean?


> 
> The test added is very good. Probably writing the test took longer than the fix.

Well, not really, because I've got sidetracked by an attempt to fix the priority logic inside the map item `QgsLayoutItemMap::layersToRender()` and I lost a lot of time there before giving up (I'm still convinced that the priority should be different but though).


> 
> Thanks @elpaso

You're welcome, thank you for testing this PR :)

File: python/core/auto_generated/layout/qgslayoutitemmap.sip.in (hunk 1 above)
Comment (elpaso) [reply python/core/auto_generated/layout/qgslayoutitemmap.sip.in:512]: @nyalldawson gentle ping: if this is a blocker I can remove the change.

Comment (nyalldawson) [reply python/core/auto_generated/layout/qgslayoutitemmap.sip.in:512]: Sorry, slipped through the cracks!

"If empty, the current project layers or thelayer specified in the map theme"

This is incorrect -- it should be something like "If the map theme is set then this list will be ignored". The map theme overrides anything specified in a setLayers call.

**Review by pblottiere — APPROVED** [review]

**Comment by pblottiere** [comment]
@elpaso Thanks, code looks good to me :+1:.

Can I push the green button or do you have some more polishing to do?

**Comment by elpaso** [comment]
> @elpaso Thanks, code looks good to me +1.
> 
> Can I push the green button or do you have some more polishing to do?

Thank you for the review!

**Pull request merged**